add intro (#3)
DevExpress-Examples/winforms-charts-sort-stacked-bars-by-total-values-with-qualitativescalecomparer: Readme.md

<!-- default badges list -->
[![](https://img.shields.io/badge/Open_in_DevExpress_Support_Center-FF7200?style=flat-square&logo=DevExpress&logoColor=white)](https://supportcenter.devexpress.com/ticket/details/T585190)
[![](https://img.shields.io/badge/📖_How_to_use_DevExpress_Examples-e9f6fc?style=flat-square)](https://docs.devexpress.com/GeneralInformation/403183)
[![](https://img.shields.io/badge/💬_Leave_Feedback-feecdd?style=flat-square)](#does-this-example-address-your-development-requirementsobjectives)
<!-- default badges end -->

> # Chart for WinForms - Sort Stacked Bars by Total Values using QualitativeScaleComparer

Author: ```suggestion
# WinForms Chart - Sort Stacked Bars by Total Values using QualitativeScaleComparer
```

The example sorts X-axis data by totals in a WinForms Chart.

> In this example, X-axis displays qualitative values. These values do not have inherent numeric order, they are plotted in the same order as series points in the collection. The example assigns a custom comparer to the [AxisBase.QualitativeScaleComparer](https://docs.devexpress.com/CoreLibraries/DevExpress.XtraCharts.AxisBase.QualitativeScaleComparer) property to sort string values in a custom order: the comparer calculates the total value for each stacked bar and sorts stacked bars by aggregated values.    

Reviewer: ```suggestion
In this example, the X-axis displays qualitative values. These values do not have an inherent numeric order - they are plotted in the same order as series points in the collection. The example assigns a custom comparer to the [AxisBase.QualitativeScaleComparer](https://docs.devexpress.com/CoreLibraries/DevExpress.XtraCharts.AxisBase.QualitativeScaleComparer) property to sort string values in a custom order: the comparer calculates the total value for each stacked bar and sorts stacked bars by aggregated values.    

```

![Chart - Sorted X-axis by totals](image/chart-sorted.png)

## Implementation Details

> This example binds the `ChartControl` to a data source created in code. It generated three series, each with ten arguments and random values. To sort stacked bars by their aggregate values (totals), do the following steps.

Reviewer: ```suggestion
This example binds the `ChartControl` to a data source created in code. It generates three series, each with ten arguments and random values. To sort stacked bars by their aggregate values (totals), do the following:

```

### Create a Comparer

Create a `ArgumentByTotalComparer` class (based on [IComparer](https://learn.microsoft.com/en-us/dotnet/api/system.collections.icomparer)). The comparer sorts arguments based on their total values stored in a dictionary:

```cs
class ArgumentByTotalComparer : IComparer {
    Dictionary<string, double> argTotalDict;

    public ArgumentByTotalComparer(Dictionary<string, double> argTotalDict) {
        this.argTotalDict = argTotalDict;
    }
    public int Compare(object x, object y) {
        return argTotalDict[(string)x].CompareTo(argTotalDict[(string)y]);
    }
}
```

### Calculate Totals for Stacked Bars

The `GetTotalByArg` method calculates the sum (the total) for a stacked bar.

```cs
double GetTotalByArg(object arg) {
    double total = 0;
    foreach (Series series in chartControl1.Series)
        foreach (SeriesPoint point in series.Points)
            if (Equals(point.Argument, arg))
                total += point.Values[0];
    return total;
}
```

### Sort Stacked Bars by Totals

> Handle the [ChartControl.BoundDataChanged](https://docs.devexpress.com/WindowsForms/DevExpress.XtraCharts.ChartControl.BoundDataChanged) event. This event fired after the chart is bound to the data source and generates series points. In the event handler you can calculate totals and other aggregations based on the loaded chart data.

Reviewer: ```suggestion
Handle the [ChartControl.BoundDataChanged](https://docs.devexpress.com/WindowsForms/DevExpress.XtraCharts.ChartControl.BoundDataChanged) event. This event fires after the chart is bound to the data source and generates series points. In the event handler, you can calculate totals and other aggregations based on the loaded chart data.

```
